crates/sui-core/src/authority_active/gossip/mod.rs

```diff
@@ -281,26 +256,18 @@ impl GossipDigestHandler {
         Self {}
     }
 
-    async fn process_response<A>(
-        aggregator: Arc<AuthorityAggregator<A>>,
+    async fn process_response(
         state: Arc<AuthorityState>,
         peer_name: AuthorityName,
         response: TransactionInfoResponse,
-    ) -> Result<(), SuiError>
-    where
-        A: AuthorityAPI + Send + Sync + 'static + Clone,
-    {
+    ) -> Result<(), SuiError> {
         if let Some(certificate) = response.certified_transaction {
-            // Process the certificate from one authority to ourselves
-            aggregator
-                .sync_authority_source_to_destination(
-                    certificate,
-                    peer_name,
-                    &LocalCertificateHandler {
-                        state: state.clone(),
-                    },
-                )
-                .await?;
+            let digest = *certificate.digest();
+            state
+                .database
+                .add_pending_certificates(vec![(digest, Some(certificate))])
```

Comment:
Here we have the option to not provide a certificate and I presume that then the cert is downloaded by the pending execution framework. Is it worth downloading the cert as part of the gossip logic (see line 302) or should we not download the cert, and let the pending execution framework download it?

My main concern is: if we do not sufficiently delay checking whether a cert is present / effects are present we may be downloading a certificate N times one for each peer we follow.

Reply:
You are correct, we may be doing redundant work here, though its no worse than before.

The issue is that at this point, we do not know if the digest corresponds to a real certificate (it could have been made up by a byzantine peer), so we can't send it to the execution driver without a cert. It is only ok to provide a digest without a cert in situations where you know the digest is real (e.g. it came from a checkpoint, or you heard the digest from f+1 peers).

So, the correct fix is to pass along to the execution driver the peer that is claiming to have the cert. Then the execution driver has a way of determining whether its real or not - if the originating peer disavows it, it can be discarded, just as we do now.

However, a more expedient fix right now might be to just add a random delay before the `effects_exists()` call